File: backend/schemas/staff_schedule.py
from __future__ import annotations

from datetime import date, datetime, time
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_serializer


class WeeklyScheduleDayRequest(BaseModel):
    day_of_week: int = Field(..., ge=0, le=6)
    start_time: str  # "HH:MM"
    end_time: str    # "HH:MM"
    is_active: bool = True


class WeeklyScheduleBulkRequest(BaseModel):
    days: list[WeeklyScheduleDayRequest]


class WeeklyScheduleDayResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    staff_id: UUID
    day_of_week: int
    start_time: time
    end_time: time
    is_active: bool

    @field_serializer("start_time", "end_time")
    def serialize_time(self, value: time) -> str:
        return value.strftime("%H:%M")


class BlockedTimeRequest(BaseModel):
    start_datetime: datetime
    end_datetime: datetime
    reason: Optional[str] = None
    block_type: str = "other"


class BlockedTimeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    staff_id: UUID
    start_datetime: datetime
    end_datetime: datetime
    reason: Optional[str] = None
    block_type: str
    created_at: datetime


class StaffAvailabilityEntry(BaseModel):
    staff_id: UUID
    first_name: str
    last_name: str
    role: str
    schedule: list[WeeklyScheduleDayResponse]


class WeeklyAvailabilityResponse(BaseModel):
    week_start: str  # YYYY-MM-DD
    staff: list[StaffAvailabilityEntry]


class ClockInResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    staff_id: UUID
    clock_in_at: datetime
    clock_out_at: Optional[datetime] = None
    date: date


class ClockOutResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    staff_id: UUID
    clock_in_at: datetime
    clock_out_at: datetime
    date: date
    total_minutes: int


class ClockStatusResponse(BaseModel):
    clocked_in: bool
    clock_in_at: Optional[datetime] = None


class AttendanceRecord(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    staff_id: UUID
    first_name: str
    last_name: str
    date: date
    clock_in_at: datetime
    clock_out_at: Optional[datetime] = None
    total_minutes: Optional[int] = None


class AttendanceSummary(BaseModel):
    staff_id: UUID
    full_name: str
    period_start: date
    period_end: date
    total_minutes: int
    records: list[AttendanceRecord]

File: types/staffSchedule.ts
export type BlockType = "lunch" | "holiday" | "personal" | "other";

export interface WeeklyScheduleDay {
  id: string;
  staffId: string;
  dayOfWeek: number; // 0=Mon..6=Sun
  startTime: string; // "HH:MM"
  endTime: string;
  isActive: boolean;
}

export interface BlockedTime {
  id: string;
  staffId: string;
  startDatetime: string; // ISO 8601
  endDatetime: string;
  reason: string | null;
  blockType: BlockType;
  createdAt: string;
}

export interface StaffAvailabilityEntry {
  staffId: string;
  firstName: string;
  lastName: string;
  role: string;
  schedule: WeeklyScheduleDay[];
  blocks: BlockedTime[];
}

export interface WeeklyAvailabilityResponse {
  weekStart: string;
  staff: StaffAvailabilityEntry[];
}

export interface ClockStatus {
  clockedIn: boolean;
  clockInAt: string | null;
}

export interface AttendanceRecord {
  id: string;
  staffId: string;
  firstName: string;
  lastName: string;
  date: string; // YYYY-MM-DD
  clockInAt: string;
  clockOutAt: string | null;
  totalMinutes: number | null;
}

export interface AttendanceSummary {
  staffId: string;
  fullName: string;
  periodStart: string;
  periodEnd: string;
  totalMinutes: number;
  records: AttendanceRecord[];
}

// --- snake_case → camelCase transforms (explicit, not generic) ---
// Explicit transforms per MEMORY.md feedback_camelizekeys_nested rule —
// generic recursive camelize breaks nested JSONB domain keys.

type RawSchedule = {
  id: string;
  staff_id: string;
  day_of_week: number;
  start_time: string;
  end_time: string;
  is_active: boolean;
};

export function camelizeSchedule(r: RawSchedule): WeeklyScheduleDay {
  return {
    id: r.id,
    staffId: r.staff_id,
    dayOfWeek: r.day_of_week,
    startTime: r.start_time,
    endTime: r.end_time,
    isActive: r.is_active,
  };
}

type RawBlockedTime = {
  id: string;
  staff_id: string;
  start_datetime: string;
  end_datetime: string;
  reason: string | null;
  block_type: BlockType;
  created_at: string;
};

export function camelizeBlockedTime(r: RawBlockedTime): BlockedTime {
  return {
    id: r.id,
    staffId: r.staff_id,
    startDatetime: r.start_datetime,
    endDatetime: r.end_datetime,
    reason: r.reason,
    blockType: r.block_type,
    createdAt: r.created_at,
  };
}

type RawAttendance = {
  id: string;
  staff_id: string;
  first_name: string;
  last_name: string;
  date: string;
  clock_in_at: string;
  clock_out_at: string | null;
  total_minutes: number | null;
};

export function camelizeAttendance(r: RawAttendance): AttendanceRecord {
  return {
    id: r.id,
    staffId: r.staff_id,
    firstName: r.first_name,
    lastName: r.last_name,
    date: r.date,
    clockInAt: r.clock_in_at,
    clockOutAt: r.clock_out_at,
    totalMinutes: r.total_minutes,
  };
}

type RawClockStatus = {
  clocked_in: boolean;
  clock_in_at: string | null;
};

export function camelizeClockStatus(r: RawClockStatus): ClockStatus {
  return {
    clockedIn: r.clocked_in,
    clockInAt: r.clock_in_at,
  };
}
